api/apiclient_test.go

```diff
@@ -153,14 +154,91 @@ func (s *apiclientSuite) TestServerRoot(c *gc.C) {
 }
 
 func (s *apiclientSuite) TestDialWebsocketStopped(c *gc.C) {
+	f := (*api.NewWebsocketDialer)(nil, api.DialOpts{})
 	stopped := make(chan struct{})
-	f := api.NewWebsocketDialer(nil, api.DialOpts{})
 	close(stopped)
 	result, err := f(stopped)
 	c.Assert(err, gc.Equals, parallel.ErrStopped)
 	c.Assert(result, gc.IsNil)
 }
 
+func (s *apiclientSuite) TestOpenWithSNIHostNameEmptyCert(c *gc.C) {
+	info := &api.Info{
+		Addrs:       []string{"foo.com:1234", "0.1.2.3:1234"},
+		SNIHostName: "foo.com",
+		SkipLogin:   true,
+	}
+	s.testSNIHostName(c, info, []apiDialInfo{{
+		location:   "wss://foo.com:1234/api",
+		hasRootCAs: false,
+		serverName: "foo.com",
+	}, {
+		location:   "wss://0.1.2.3:1234/api",
+		hasRootCAs: false,
+		serverName: "foo.com",
+	}})
+}
+
+func (s *apiclientSuite) TestOpenWithSNIHostNameWithCACert(c *gc.C) {
+	info := &api.Info{
+		Addrs:       []string{"foo.com:1234", "0.1.2.3:1234"},
+		SNIHostName: "foo.com",
+		SkipLogin:   true,
+		CACert:      jtesting.CACert,
+	}
+	s.testSNIHostName(c, info, []apiDialInfo{{
+		location:   "wss://foo.com:1234/api",
+		hasRootCAs: true,
+		serverName: "juju-apiserver",
+	}, {
+		location:   "wss://0.1.2.3:1234/api",
+		hasRootCAs: true,
+		serverName: "juju-apiserver",
+	}})
+}
+
+type apiDialInfo struct {
+	location   string
+	hasRootCAs bool
+	serverName string
+}
+
+// testSNIHostName tests that when the API is dialed with the given info,
+// api.newWebsocketDialer is called with the expected information
+// (one element for each call to newWebsocketDialer)
+func (s *apiclientSuite) testSNIHostName(c *gc.C, info *api.Info, expectDials []apiDialInfo) {
+	dialed := make(chan *websocket.Config)
+	s.PatchValue(api.NewWebsocketDialer, func(cfg *websocket.Config, opts api.DialOpts) func(<-chan struct{}) (io.Closer, error) {
+		return func(<-chan struct{}) (io.Closer, error) {
+			dialed <- cfg
+			return nil, errors.New("nope")
+		}
+	})
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		conn, err := api.Open(info, api.DialOpts{})
+		c.Check(conn, gc.Equals, nil)
+		c.Check(err, gc.ErrorMatches, `nope`)
+	}()
+	for _, expect := range expectDials {
```

Comment:
I think we can make this test a little easier to read through by inverting what you're doing here:

- Create a timeout channel that your fake `NewWebsocketDialer` can close over (or somehow use the one we're ignoring on L212 -- actually, why isn't that bubbling up to the API we're interacting with?).
- In your anonymous goroutine, do sleep for jtesting.LongWait and then close the timeout channel (or maybe just fail the test outright?). This is semantically different in that we have one timeout for all sends, but `10*time.Second` should be long enough for all sends.
- Invert this for loop to range over `dialed`

This way I think you can get rid of both `select` blocks and we're looking at a simple for loop.